### DISCUSSION TITLE
please help!onAdd and onRemove unable to trigger,but onStateChange can trigger

### DISCUSSION THREAD

**Discussion by a346573450** [colyseus/colyseus#787] [Q&A]
### **Room.onStateChange can trigger ,but onAdd and onRemove Unable to trigger. No network requests received.
This problem has been bothering me for many days.
Can you give me some help? I hope to receive your guidance.thanks!**

**### this is client code**
_**pic1**_
![1](https://github.com/user-attachments/assets/f64515d1-18fd-46b8-834a-db3bc88138b3)

_**pic2**_
![2](https://github.com/user-attachments/assets/c2eec748-803c-42fa-a4be-0b3513120639)

### **this is server code**
_**pic1**_
![3](https://github.com/user-attachments/assets/6adbe435-791f-4484-ae69-2e7d1baa1494)

_**pic2**_
![4](https://github.com/user-attachments/assets/4227c96c-d8f2-4c15-a73e-b2fd8c4e82ef)

_**pic3**_
![5](https://github.com/user-attachments/assets/db1050ae-1393-4a7a-8a89-eb9d5c9652a4)

**Comment by endel** [comment]
Hi @a346573450, please take a look at the documentation here: https://docs.colyseus.io/state/schema-callbacks/#onadd-fn-item-key-triggerall-true 

You seem to be using 0.14 callbacks (`.onAdd = ...`, `.onChange = ...`). 
The way to attach callbacks have changed on 0.15 to use callbacks instead (`.onAdd(...)`, `.onChange(...)`)

I hope this helps! Cheers!

**Reply by a346573450** [reply]
thanks i love you!
